### PR TITLE
Add test case for PositionedCryptoInputStream

### PR DESCRIPTION
**Pull request by colinmjj**

### CODE REVIEW

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+  }
+
+  private void prepareData() throws IOException {
+    Cipher cipher = null;
+    try {
+      cipher = (Cipher)ReflectionUtils.newInstance(
+              ReflectionUtils.getClassByName(jceCipherClass), props, transformation);
+    } catch (ClassNotFoundException cnfe) {
+      throw new IOException("Illegal crypto cipher!");
+    }
```
Comment (winningsix) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: Hint: You may use ```private Cipher getCipher(String cipherClass) throws IOException``` in this class.

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    out.close();
+    encData = baos.toByteArray();
+
+    // create the file with the encryption data for the test
+    try {
+      testDataFilepath = Files.createTempFile("encData", ".tmp");
+      SeekableByteChannel sbc = Files.newByteChannel(testDataFilepath, StandardOpenOption.APPEND);
+      ByteBuffer bfSrc = ByteBuffer.wrap(encData);
+      sbc.write(bfSrc);
+      sbc.close();
```
Comment (winningsix) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: Please do the close operation in the finally block instead.

**Comment by jerrychenhf** [comment]
Thanks @colinmjj for contributing the patch.
We add PositionedChannelInput, there are several problems.
1. first SeekableByteChannel is only available at JDK7 which makes the library only usable by JDK7 or later. We recently removed the constraint.
2. The PositionedChannelInput is NOT possibly implemented by SeekableByteChannel. if there is no position read method provided. positioned read is different from seek. positioned read means it can keep multiple read on different position but not interference each other.
3. Trivially, readFully is implemented wrong.

Overally, the patch code somewhat deviated from the target of "add test case for PositionedCryptoInputStream". I suggest that we can write a test Input which support positioned read based on memory buffers and simply use it to test PositionedChannelInput.

**Comment by colinmjj** [comment]
@jerrychenhf , thanks for the comments. I'll write the test Input and test PositionedCryptoInputStream with it.

**Comment by jerrychenhf** [comment]
Per latest discussion in the Apache Commons community, we don't have to stick on JDK6. JDK7 or later is OK for us. While SeekableByteChannel  is still not the right option here:)

**Comment by colinmjj** [comment]
@jerrychenhf , @winningsix , thanks for the comments, the patch is updated for review.

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+  private Cipher getCipher(String cipherClass) throws IOException {
+    try {
+      return (Cipher)ReflectionUtils.newInstance(
+              ReflectionUtils.getClassByName(cipherClass), props, transformation);
+    } catch (ClassNotFoundException cnfe) {
+      throw new IOException("Illegal crypto cipher!");
+    }
+  }
+
+  class PositionedInputForTest implements Input {
```
Comment (jerrychenhf) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: There is a critical problem in the PositionedInputForTest implementation.
byte[] buf; 
The member "buf" was used in a confusing way.  I think you want to use as byte array input data.  But getBuf method use the same member to store a temp buffer. and in the read(ByteBuffer) method, it read data from buf and copy into the same buf to a different pos.

You may mean to:
byte[] data;  
long pos;  
long count;  

byte[] buf;
int bufferSize;  

A more important questions to ask:
Since the PositionedInputForTest is totally behavior wrong. Does the test cases still pass? Does the test cases designed valid?

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+        read += copyLen;
+        offset += copyLen;
+        pos += copyLen;
+        remaining -= copyLen;
+      }
+      return read;
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
```
Comment (jerrychenhf) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: The logic in the method can be improved. It is currently difficult to understand and hard to follow whether it is correct or not. What if it is as following:
if (n <= 0) {
  return 0;
}

long remaining = count - pos;
if(remaining < n) {
  n = remaining;
}
pos += n;

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        if (nbytes < 0) {
+          throw new EOFException("End of file reached before reading fully.");
+        }
+        nread += nbytes;
+      }
+    }
+
+    @Override
+    public void seek(long position) throws IOException {
+      if (position >= 0 && position < count) {
```
Comment (jerrychenhf) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: 1. position should be able to equal count which means the end of the stream. The current logic will never be able to seek to the end.
2. Shall we throw an exception if the position is invalid, for example negative case?

**Comment by jerrychenhf** [comment]
Thanks @colinmjj for updating the patch. 
I have reviewed and made some comments. And there are some critical bugs there. You may need to check the test cases for whether the test cases are properly constructed.

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java (hunk 3 above)
Comment (colinmjj) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: Critical catch, actually, there is no test case for read(ByteBuffer dst), so it wasn't exposed during the test.
I'll also add test cases for read(ByteBuffer dst).

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java (hunk 4 above)
Comment (colinmjj) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: Thanks for the comments, will update the logic.

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+    public PositionedInputForTest(byte[] data, int bufferSize) {
+      this.data = data;
+      this.pos = 0;
+      this.count = data.length;
+      this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public int read(ByteBuffer dst) throws IOException {
```
Comment (jerrychenhf) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: I see this code is from other classes. But for read(ByteBuffer dst) implementation here, it is not necessary do it in a complex way like this. Temp buffer is not necessary
Since it is backed by a byte array underlayer, why not simply check and do a copy once like, what you do in positioned read (or actually you can actually call directly the positioned read and increment the pos according to the return value.)

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+    OutputStream out = new CryptoOutputStream(baos, cipher, defaultBufferSize, key, iv);
+    out.write(data);
+    out.flush();
+    out.close();
+    encData = baos.toByteArray();
+  }
+
+  public void setUp() throws IOException {}
+
+  private PositionedCryptoInputStream getCryptoInputStream(Cipher cipher, int bufferSize,byte[] iv)
```
Comment (jerrychenhf) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: A question here: Is there any reason that iv is passed as parameter but key is not?  I think they should be consistent either both be passed as parameter or both not.

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    PositionedCryptoInputStream in = getCryptoInputStream(getCipher(cipherClass), defaultBufferSize, iv);
+    in.seek(position);
+    ByteBuffer buf = ByteBuffer.allocate(length);
+    int n = in.read(buf);
+    Assert.assertEquals(exceptedLength, n);
+    byteArrayCompare(n, buf.array(), position, length);
+    in.close();
+  }
+
+  private void positionedReadCheck(String cipherClass, int position, int offset, int length) throws Exception {
```
Comment (jerrychenhf) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: For the positionedReadCheckmethods, I would suggest to use while loop to read multiple times and check until to the end of stream. The size of each read can be configured so that we can use different size to do a multiple tests.

If the other method can be done in the same way, we can do that.

**Comment by jerrychenhf** [comment]
Thanks @colinmjj for updating the patch.

I made a few of further comments. And one additional suggest is we need to test the case that when there are multiple positioned read actions and there is one read action and they are not interfere each other.

**Comment by colinmjj** [comment]
@jerrychenhf , thanks for your all comments, and the patch is updated according to the comments.

File: src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+
+    public PositionedInputForTest(byte[] data) {
+      this.data = data;
+      this.pos = 0;
+      this.count = data.length;
+    }
+
+    @Override
+    public int read(ByteBuffer dst) throws IOException {
+      int remaining = dst.remaining();
```
Comment (jerrychenhf) [inline src/test/java/com/intel/chimera/stream/PositionedCryptoInputStreamTest.java]: The wrong usage of available method even the result is correct here.

Although I previously suggest we may call positioned. But it is actually more complex than we simple not call it.The read method can be as simple as:
int remaining = count - pos;
if(remaining <= 0) {
  return -1;
}

 int length = Math.min(dst.remaining(), remaining); 
if(length == 0) {
  return 0;
}
dst.put(data, pos, length);
pos += length;
return length;

**Comment by jerrychenhf** [comment]
Thanks @colinmjj. Ready to go after addressing my last comment on the read(ByteBuffer) method.

**Comment by colinmjj** [comment]
@jerrychenhf , patch is updated, thanks for the review.

**Comment by jerrychenhf** [comment]
Thanks @colinmjj for contributing!
LGTM.
